src/gamestate.rs

```diff
@@ -1,795 +0,0 @@
-        self.players_turn = if self.players_turn == Color::Black {
-            Color::White
-        } else {
-            Color::Black
-        };
-    }
-
-    fn all_valid_pieces_moves(&self, color: Color) -> HashSet<Point> {
-        let mut moves = HashSet::new();
-        let pieces = if color == Color::Black {
-            &self.black_pieces
-        } else {
-            &self.white_pieces
-        };
-        for (x, y) in pieces {
-            if let Tile::Piece(piece) = self.board[*x as usize][*y as usize] {
-                moves.extend(self.valid_piece_moves(&MovingPiece::new(piece, (*x, *y))))
-            }
-        }
-        moves
-    }
-
-    fn all_pieces_moves(&self, color: Color) -> HashSet<Point> {
-        let mut moves = HashSet::new();
-        let pieces = if color == Color::Black {
-            &self.black_pieces
-        } else {
-            &self.white_pieces
-        };
-        for (x, y) in pieces {
-            if let Tile::Piece(piece) = self.board[*x as usize][*y as usize] {
-                moves.extend(self.all_piece_moves(&MovingPiece::new(piece, (*x, *y))))
-            }
-        }
-        moves
-    }
-
-    fn valid_piece_moves(&self, selected_piece: &MovingPiece) -> HashSet<Point> {
-        let moves = self.all_piece_moves(selected_piece);
-        moves
-            .into_iter()
-            .filter(|x| {
-                let selected_move = Move::new(selected_piece.point.to_owned(), x.to_owned());
-                self.is_valid_move(selected_move)
-            })
-            .collect()
-    }
-
-    fn all_piece_moves(&self, selected_piece: &MovingPiece) -> HashSet<Point> {
-        let mut moves = HashSet::new();
-        moves.extend(&self.piece_moves(selected_piece));
-        moves.extend(&self.piece_attack_moves(selected_piece));
-        moves
-    }
-
-    fn piece_moves(&self, selected_piece: &MovingPiece) -> HashSet<Point> {
-        use PieceState::*;
-        match selected_piece.piece.state {
-            King => self.king_moves(selected_piece.point),
-            Queen => self.queen_moves(selected_piece.point),
-            Rook => self.rook_moves(selected_piece.point),
-            Bishop => self.bishop_moves(selected_piece.point),
-            Knight => self.knight_moves(selected_piece.point),
-            Pawn => self.pawn_moves(selected_piece.point),
-        }
-    }
-
-    fn piece_attack_moves(&self, selected_piece: &MovingPiece) -> HashSet<Point> {
-        use PieceState::*;
-        match selected_piece.piece.state {
-            King => self.attack_king_moves(selected_piece.point),
-            Queen => self.attack_queen_moves(selected_piece.point),
-            Rook => self.attack_rook_moves(selected_piece.point),
-            Bishop => self.attack_bishop_moves(selected_piece.point),
-            Knight => self.attack_knight_moves(selected_piece.point),
-            Pawn => self.attack_pawn_moves(selected_piece.point),
-        }
-    }
```

Comment:
consider refactoring to add a method to the enum variant itself. 
It is a strange choice to have the game state object act as a god object for all the other pieces. 
you should think about how to separate the logic so that each piece can move as long as it implements the Move trait

Reply:
ah like the move method would pass in its point and the current board also whats the point of the trait if theres only one type

Reply:
I meant have a trait for a game piece, each game piece has a different implementation of move but can share base logic

Reply:
also for enpassent for pawns i needed the previous board as well